[1.20.6] Redstone updates and gametests structure fixes

Jonathing opened the pull request:
- Backport of #10391 to 1.20.6.
- Backport of #10392 to 1.20.6.
- Backport of e69b07a to 1.20.6.
- Partially fixes #9973 for 1.20.6.

> [!IMPORTANT]
> **The new update order is gated behind a Forge server config.** The host world must have this config enabled to see the new update order take effect. This is to prevent breaking changes with redstone and neighbor updates since Forge for this version is already stable.

> [!CAUTION]
> **This PR includes a breaking change to game tests.** Due to the nature of needing to test rotations, fix has been included make block position assertions behave identical to 1.21.3. This fix can be disabled in Forge's common config, and is included in there and not the server config so that development environments do not need to set the config value on every world, or arbitrarily make a `defaultconfigs` entry. All 40 of Forge's own game tests still pass with this change.

LexManos committed:
Fix game test structure rotations being inconsistant (#10391)

LexManos committed:
Fix redstone update order not matching vanilla. (#10392)

LexManos committed:
Fix redstone updates for comparators going up/down incorrectly. Related: #9973

Jonathing committed:
Backport to 1.21.1

Jonathing committed:
Make block pos assertions behave like in 1.21.3

!!! THIS CAN BE DISABLED IN FORGE'S COMMON CONFIG !!!

Jonathing committed:
Backport to 1.20.6

Jonathing committed:
Gate fixed update order behind server config

Jonathing committed:
Add language entries for the config

LexManos commented on patches/minecraft/net/minecraft/core/Direction.java.patch:
```diff
@@ -0,0 +1,14 @@
++    // In vanilla updates are notified only on the horizontal plane, in Forge we also notify up and down
++    private static final Direction[] UPDATE_ORDER = Stream.concat(Plane.HORIZONTAL.stream(), Plane.VERTICAL.stream()).toArray(Direction[]::new);
++    public static final Direction[] getUpdateOrder() {
++        return UPDATE_ORDER.clone();
```
Oh wow, so this exposes a stupid decision made ages ago.
This needs to be a list or iterable not an array. Because we don't want to keep cloneing the array every time it's called.
So keep the change from future versions where it's a List

Jonathing replied:
I made it an array because I couldn't use `?:` in the Level patch otherwise. How should I deal with that? Or should I backport the values view?

Jonathing replied:
Or should I backport the values view? Just thought of that.

LexManos commented on patches/minecraft/net/minecraft/world/level/Level.java.patch:
```diff
@@ -206,9 +207,13 @@
 -                    if (blockstate.is(Blocks.COMPARATOR)) {
 +                    if (blockstate.getWeakChanges(this, blockpos)) {
++                        // Forge: Replace Level#neighborChanged with BlockState#onNeighborChange. Boolean.valueOf(false) is used to account for problematic mixins (see #10402).
++                        if (Boolean.valueOf(false))
```
We don't need this

Jonathing replied:
Ok, will remove later.

Jonathing replied:
I just tested this change on 1.20.1 on some big modpacks. The config value was off, but if there was a Mixin issue, the game would have crashed. So yeah, this is fine. The weird one in 1.20.1 for villager types was just an unfortunate case.

Jonathing committed:
Don't use Boolean.valueOf(false) garbage

Jonathing committed:
Wrap views of directions in lists

`Stream#toList` uses `Collections#unmodifiableList`, so no need to use
it for `UPDATE_ORDER`

LexManos commented on patches/minecraft/net/minecraft/world/level/Level.java.patch:
```diff
@@ -206,6 +207,9 @@
++                        // Forge: Use IForgeBlockState#onNeighborChange, the default impl redirects to Level#neighborChanged
++                        blockstate.onNeighborChange(this, blockpos, p_46718_);
++                        if (false)
                          this.neighborChanged(blockstate, blockpos, p_46719_, p_46718_, false);
```
Just delete the line.. no need for if(false) we're deleting a line and replacing it with like we do in modern versions.

Jonathing replied:
Ok, I wasn't sure since `if (false)` is used in other places. Noted that full-on replacements to just delete the original line.

LexManos replied:
`if (false)` is used to keep the patch small. If we are nooping large branches then we use `if (false)` but if its jsut one line then its  just deleted to be more clear.

Jonathing committed:
Remove unnecessary if (false)
